sync/metrics.go

```diff
@@ -24,27 +39,99 @@ func newMetrics() (*metrics, error) {
 		return nil, err
 	}
 
+	syncLoopStarted, err := meter.Int64Counter("sync_loop_started", metric.WithDescription("sync loop started"))
+	if err != nil {
+		return nil, err
+	}
+
+	trustedPeersOutOfSync, err := meter.Int64Counter("tr_peers_out_of_sync", metric.WithDescription("trusted peers out of sync"))
+	if err != nil {
+		return nil, err
+	}
+
+	laggingHeadersStart, err := meter.Int64Counter("sync_lagging_hdr_start", metric.WithDescription("lagging header start"))
+	if err != nil {
+		return nil, err
+	}
+
+	subjectiveHead, err := meter.Int64ObservableGauge("sync_subjective_head", metric.WithDescription("subjective head height"))
+	if err != nil {
+		return nil, err
+	}
+
+	headerTimestamp, err := meter.Float64Histogram("sync_subjective_head_ts",
+		metric.WithDescription("subjective_head_timestamp"))
+	if err != nil {
+		return nil, err
+	}
+
 	m := &metrics{
-		totalSyncedGauge: totalSynced,
+		ctx:                   context.Background(),
+		totalSyncedGauge:      totalSynced,
+		syncLoopStarted:       syncLoopStarted,
+		trustedPeersOutOfSync: trustedPeersOutOfSync,
+		laggingHeadersStart:   laggingHeadersStart,
+		headerTimestamp:       headerTimestamp,
+		headersThreshHold:     headersThreshHold,
 	}
 
 	callback := func(ctx context.Context, observer metric.Observer) error {
 		observer.ObserveFloat64(totalSynced, float64(m.totalSynced.Load()))
+		observer.ObserveInt64(subjectiveHead, m.subjectiveHead.Load())
 		return nil
 	}
-	_, err = meter.RegisterCallback(callback, totalSynced)
+
+	_, err = meter.RegisterCallback(callback, totalSynced, subjectiveHead)
 	if err != nil {
 		return nil, err
 	}
-
 	return m, nil
 }
 
-// recordTotalSynced records the total amount of synced headers.
 func (m *metrics) recordTotalSynced(totalSynced int) {
 	if m == nil {
 		return
 	}
-
 	m.totalSynced.Add(int64(totalSynced))
 }
+
+func (m *metrics) recordSyncLoopStarted() {
```

Comment:
pass context in here vs setting on the struct?

Reply:
we can pass syncer's context, but in case syncer is stopped(and thus context is canceled) we won't set the metric.

Reply:
this is whyt this pattern exists and we should follow it here as well!
```
func (m *subscriberMetrics) observe(ctx context.Context, observeFn func(context.Context)) {
	if m == nil {
		return
	}
	if ctx.Err() != nil {
		ctx = context.Background()
	}

	observeFn(ctx)
}
```

Reply:
So pass the contexts properly